msg.go

```diff
@@ -624,18 +624,25 @@ func UnpackRRWithHeader(h RR_Header, msg []byte, off int) (rr RR, off1 int, err
 		rr = &RFC3597{Hdr: h}
 	}
 
-	if noRdata(h) {
-		return rr, off, nil
+	if off < 0 {
+		return nil, off, &Error{err: "off must be positive"}
 	}
 
 	end := off + int(h.Rdlength)
+	if end < off || end > len(msg) {
+		return nil, end, &Error{err: "bad rdlength"}
+	}
+
+	if noRdata(h) {
+		return rr, off, nil
+	}
 
 	off, err = rr.unpack(msg, off)
 	if err != nil {
 		return nil, end, err
 	}
 	if off != end {
-		return &h, end, &Error{err: "bad rdlength"}
+		return nil, end, &Error{err: "bad rdlength"}
```

Comment:
I think the idea was to leave the header returned so the caller could still do something with it. This is also probable part of the assumed API of this lib nowdays

Reply:
I've reverted the changes to `UnpackRRWithHeader` and will propose them separately.